telomerecat/telbam.py

```diff
@@ -0,0 +1,134 @@
+from typing import List
+import os
+import sys
+import pysam
+import logging
+import tempfile
+import multiprocessing as mp
+import subprocess
+from functools import partial
+
+from telomerecat.constants import TEL_PATS, HTS_EXT_TO_AF_MODE
+from pysam import AlignmentFile
+
+def _log_setup(loglevel):
+    logging.basicConfig(level=getattr(logging, loglevel.upper()), format='%(levelname)s: %(message)s')
+
+
+def get_align_file(path:str, mode='r', template=None, expectIndex=True, threads=1):
+    hts_ext = os.path.splitext(path)[-1]
+    xam_type = HTS_EXT_TO_AF_MODE[hts_ext]
+    if expectIndex is False:
+        save = pysam.set_verbosity(0)
+    af = pysam.AlignmentFile(path, f'{mode}{xam_type}', template=template, threads=threads)
+    if expectIndex is False:
+        pysam.set_verbosity(save)
+    return af
+
+
+def get_processes(processes):
+    if processes == -1:
+        processes = len(os.sched_getaffinity(0))
+    return processes
+
+
+def get_hts_processes(processes):
+    if processes > 2:
+        processes = 2
+    return processes
+
+
+def telbam_path(dest_dir:str, xam_file:str):
+    (base, ext) = os.path.splitext(os.path.basename(xam_file))
+    return os.path.join(dest_dir, f'{base}_telbam.bam')
+
+
+def collate_pairs(xam_file: str, tmpdir:str, processes=1, reference=None):
+    pairs = os.path.join(tmpdir, 'pairs.bam')
+    collate_opts = ['--no-PG', '-f', '-n', '1', '-l', '1', '-r', '5000000']  # ~8GB RAM
+    if processes > 1:
+        collate_opts.extend(['-@', str(processes)])
+    if reference:
+        collate_opts.extend(['--reference', reference])
+    collate_opts.extend(['-o', pairs, xam_file, os.path.join(tmpdir, 'collate')])
+    pysam.collate(*collate_opts) # don't forget to splat, this isn't perl
+    return pairs
+
+
+def pairs_to_telbam(af_pairs:AlignmentFile, af_telbam:AlignmentFile):
+    read_iter = af_pairs.fetch(until_eof=True)
+    while True:
+        read_a = next(read_iter, None)
+        if read_a is None:
+            break
+        read_b = next(read_iter)
+        qseq = read_a.query_sequence
+        if TEL_PATS[0] in qseq or TEL_PATS[1] in qseq:
+            af_telbam.write(read_a)
+            af_telbam.write(read_b)
+        else:
+            qseq = read_b.query_sequence
+            if TEL_PATS[0] in qseq or TEL_PATS[1] in qseq:
+                af_telbam.write(read_a)
+                af_telbam.write(read_b)
+    return
+
+
+def to_telbam(xam_file:str, outbam_dir:str, tmpdir:str, quick=False, hts_processes=1, reference=None):
+    telbam = telbam_path(outbam_dir, xam_file)
```

Comment:
`quick` is unused.

Reply:
Ah, yes.  leftover from when I was using a pre-collated BAM as input for testing.